iroh/src/baomap/mem.rs

```diff
@@ -556,6 +556,14 @@ impl baomap::Store for Store {
     }
 
     fn temp_tag(&self, tag: HashAndFormat) -> TempTag {
+        let mut state = self.0.state.write().unwrap();
+        // add to live set immediately. This prevents this blob from being deleted
+        // if this happens between a mark and a sweep phase.
+        state.live.insert(tag.0);
+        // increase the ref count
+        let entry = state.temp.entry(tag).or_default();
+        // panic if we overflow an u64
+        *entry = entry.checked_add(1).unwrap();
         TempTag::new(tag, Some(self.0.clone()))
```

Comment:
So the above seems to be the same as the code in `on_clone`. But the `clone` called here is not triggering the callback, as you implemented `LivenessTracker` on `Inner`, not on `Arc<Inner`. Why do this duplication/separation?

Reply:
Sigh.

Yes, the code in on_clone does the exact same thing, except for adding the hash to the live set. I could just remove this and do the refcount increase in TempTag::new(), but then I would have to grab the mutex twice.

I think this is how I introduced this bug in the first place. I tried to DRY creation and refcount increase and forgot to copy over the live set insertion...

Any idea how to DRY this without needlessly double locking?

Reply:
The things I tried to DRY these things ended up making things worse. E.g. you could have a fn on the state (where you already have the &mut) that does the refcount increase. But is that really worth having another fn or another struct to wrap the temp table?

Reply:
I am less worried about the DRYness of the code, more about this potentially being done twice due to `Arc::clone` vs `Inner::clone`

Reply:
If you were to increase the refcount by 2 on construction, that data could never be erased because it would not go back to 0. It would immediately be caught by the tests.

But I got another idea. Maybe not have the temp tags in the live set, but just check both sets in is_live? Will try that.

Reply:
Nah, that does not work. Temp has a HashAndFormat as key, and is_live requires just a hash, so the test would have to be like this:

```
        let res = state.live.contains(hash)
            || state.temp.contains_key(&HashAndFormat::raw(*hash))
            || state.temp.contains_key(&HashAndFormat::hash_seq(*hash));
```

I mean, it works...

Reply:
wouldn't it make more sense to store the raw hash in the set and the format in the value part? inclusion checks the format shouldn't matter asfaiu

Reply:
for the temp tags it does matter if it is hashseq or raw. a temp tag with a format of hashseq protects the children too.

What you could do is to make the temp map a map from hash to (raw_count, hash_seq_count). But you need separate counts.